src/RRule.php

```diff
@@ -0,0 +1,276 @@
+    {
+        $startDate = clone $start;
+        // When the start time contains microseconds, the first recurrence will always be skipped, as
+        // the transformer operates only up to seconds level. See also the upstream issue #155
+        $startDate->setTime($start->format('H'), $start->format('i'), $start->format('s'));
+
+        $this->rrule->setStartDate($startDate);
+
+        return $this;
+    }
+
+    /**
+     * Set the end time of this frequency
+     *
+     * @param DateTimeInterface $end
+     *
+     * @return $this
+     */
+    public function endAt(DateTimeInterface $end): self
+    {
+        $this->rrule->setEndDate($end);
+
+        return $this;
+    }
+
+    /**
+     * Get the frequency of this rule
+     *
+     * @return string
+     */
+    public function getFrequency(): string
+    {
+        return $this->frequency;
+    }
+
+    /**
+     * Get a set of recurrences relative to the given time and bounded to the configured generator's limit
+     *
+     * @param DateTimeInterface $dateTime
+     * @param int $limit Limit the recurrences to be generated to the given value
+     * @param bool $include Whether to include the passed time in the result set
+     *
+     * @return Generator
+     */
+    public function getNextRecurrences(
+        DateTimeInterface $dateTime,
+        int $limit = self::DEFAULT_LIMIT,
+        bool $include = true
+    ): Generator {
+        $resetTransformerConfig = function (int $limit = self::DEFAULT_LIMIT): void {
+            $this->transformerConfig->setVirtualLimit($limit);
+            $this->transformer->setConfig($this->transformerConfig);
+        };
+
+        if ($limit > self::DEFAULT_LIMIT) {
+            $resetTransformerConfig($limit);
+        }
+
+        $constraint = new AfterConstraint($dateTime, $include);
+        if (! $this->rrule->repeatsIndefinitely()) {
+            // When accessing this method externally (not by using `getNextDue()`), the transformer may
+            // generate recurrences beyond the configured end time.
+            $end = $this->rrule->getEndDate() ?? $this->rrule->getUntil();
+            $constraint = new BetweenConstraint($dateTime, $end, $include);
+        }
+
+        // Setting the start date to a date time smaller than now causes the underlying library
+        // not to generate any recurrences when using the regular frequencies such as `MINUTELY` etc.
+        // and the `$countConstraintFailures` is set to true. We need also to tell the transformer
+        // not to count the recurrences that fail the constraint's test!
+        $recurrences = $this->transformer->transform($this->rrule, $constraint, false);
+        foreach ($recurrences as $recurrence) {
+            yield $recurrence->getStart();
+        }
+
+        if ($limit > self::DEFAULT_LIMIT) {
+            $resetTransformerConfig();
+        }
+    }
+
+    #[\ReturnTypeWillChange]
+    public function jsonSerialize()
```

Comment:
Why does this implement `JsonSerializable` but does not return valid JSON?

Please just use `__toString` or some other method. This is just wrong.